mod/file/views/default/forms/file/upload.php

```diff
@@ -33,54 +33,67 @@
 
 $upload_limit = elgg_echo('file:upload_limit', array(elgg_format_bytes($max_upload)));
 
-?>
-<div class="mbm elgg-text-help">
-	<?php echo $upload_limit; ?>
-</div>
-<div>
-	<label><?php echo $file_label; ?></label><br />
-	<?php echo elgg_view('input/file', array('name' => 'upload')); ?>
-</div>
-<div>
-	<label><?php echo elgg_echo('title'); ?></label><br />
-	<?php echo elgg_view('input/text', array('name' => 'title', 'value' => $title)); ?>
-</div>
-<div>
-	<label><?php echo elgg_echo('description'); ?></label>
-	<?php echo elgg_view('input/longtext', array('name' => 'description', 'value' => $desc)); ?>
-</div>
-<div>
-	<label><?php echo elgg_echo('tags'); ?></label>
-	<?php echo elgg_view('input/tags', array('name' => 'tags', 'value' => $tags)); ?>
-</div>
-<?php
-
-$categories = elgg_view('input/categories', $vars);
-if ($categories) {
-	echo $categories;
-}
-
-?>
-<div>
-	<label><?php echo elgg_echo('access'); ?></label><br />
-	<?php echo elgg_view('input/access', array(
+$fields = [
+	[
+		'type' => 'file',
+		'name' => 'upload',
+		'label' => $file_label,
+		'help' => $upload_limit,
+		'value' => ($guid),
+		'required' => (!$guid),
+	],
+	[
+		'type' => 'text',
+		'name' => 'title',
+		'value' => $title,
+		'label' => elgg_echo('title'),
+	],
+	[
+		'type' => 'longtext',
+		'name' => 'description',
+		'value' => $desc,
+		'label' => elgg_echo('description'),
+	],
+	[
+		'type' => 'tags',
+		'name' => 'tags',
+		'value' => $tags,
+		'label' => elgg_echo('tags'),
+	],
+	[
+		'type' => 'categories',
+	],
+	[
+		'type' => 'access',
 		'name' => 'access_id',
 		'value' => $access_id,
 		'entity' => get_entity($guid),
 		'entity_type' => 'object',
 		'entity_subtype' => 'file',
-	)); ?>
-</div>
-<div class="elgg-foot">
```

Comment:
In general, I think we should figure out a way to not include the submit button within a form view. Would be some much easier to extend forms. IMO, we could pass the submit label with form form vars.